### ISSUE TITLE
[BUG] Pure-python mode fails at handling list type

### ISSUE THREAD

**Issue by 00sapo** [cython/cython#4089]
**Describe the bug**
1. When using lists in pure-python mode, I get better results when using no typing in respect to using typing
2. Using lists in pure-python mode with C++ is slow
3. Using lists in pure-python mode with standard python annotations slow downs in respect to no typing

Here a little table with my benchmarks:

|                               | No annotations | Annotations from typing | Annotations from cython |
|-------------------------------|----------------|-------------------------|-------------------------|
| Pure Python                   | 11.22          |                         |                         |
| C++ Cython (pure-python mode) | 02.69          | 03.80                   | 05.37                   |
| C Cython (pure-python mode)   | 02.66          | 03.13                   | 08.27                   |
| C++ Cython (.pyx)             |                |                         | 02.39                   |

**To Reproduce**
Code to reproduce the behaviour:
https://github.com/00sapo/cython_list_test

**Expected behavior**
1. Cython typing fastest than no typing
2. standard python annotations faster than no typing
3. C++ fast at least as C (since it can rely on vector)

**Environment (please complete the following information):**
Tested on:

*    Intel Core i5-8250U - 64 bits - Quad Core L2 cache: 6 MiB
*   Linux 5.4.105-1-MANJARO x86_64
*   GCC 10.2.0
*   Python 3.9.2
*   Cython 0.29.17

**Comment by da-woods** [comment]
* The `typing` module is pretty much ignored by Cython (it may be used as a tiny hint for loops I think). That'll change with https://github.com/cython/cython/pull/3910 (if accepted). Annotating as `list` should work though.

* Using c++ vector in place of list requires the list to be copied into a C++ vector. That has a cost and is probably only worthwhile if you're going to keep it in vector form for use in C++ longer term.

* Too much typing can sometimes be a mistake if all you do is add type-checks and conversions to/from c.

**Comment by scoder** [comment]
Closing. `typing` support is a different ticket, and the rest should go to the mailing list.

**Comment by 00sapo** [comment]
>     * The `typing` module is pretty much ignored by Cython (it may be used as a tiny hint for loops I think). That'll change with #3910 (if accepted). Annotating as `list` should work though.

But why using it make the code slower? Should I open another ticket on this?

>     * Using c++ vector in place of list requires the list to be copied into a C++ vector. That has a cost and is probably only worthwhile if you're going to keep it in vector form for use in C++ longer term.

This doesn't happen in the example code, because lists and vectors are created from scratch


> Closing. `typing` support is a different ticket, and the rest should go to the mailing list.

It's not only about `typing` module support but also about:

* C++ incredibly slower than C 
* annotations from cython slower than no annotations.

**Comment by da-woods** [comment]
This is my last reply on this because it really should go on the cython-users mailing list:

>>    * The `typing` module is pretty much ignored by Cython (it may be used as a tiny hint for loops I think). That'll change with #3910 (if accepted). Annotating as `list` should work though.

> But why using it make the code slower? Should I open another ticket on this?

I don't believe it is making the code slower. I believe `count: float = 0` is probably making the code slower because it adds a C-Python conversion on every loop iteration. The same idea probably applies to "annotations from cython slower than no annotations.".

Have a look at the annotated html code (`cython -a`) and if you can't work out what's going wrong then please ask on the mailing list.

**Comment by 00sapo** [comment]
I added HTML files as generated by `cython -a` and nothing strange appears to me.

1. `count: float = 0` is executed only once. At each loop, it takes an object that should be a C object and sum it to a C object, so no Python-C nor C-Python conversion happens. Even if lists contain Python objects, there should be only one conversion Python-C at each loop.

2. C++ code is unexplainable slower than C.

3. Cython types are slower than Python types. 

> This is my last reply on this because it really should go on the cython-users mailing list:

It's not a bug, it's a feature

**Comment by da-woods** [comment]
I'm not very good at not being drawn into this

`count *= a_list[i][j]` is transformed to:

```
__pyx_t_4 = PyFloat_FromDouble(__pyx_v_count); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 10, __pyx_L1_error)
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_8 = __Pyx_GetItemInt(__pyx_v_a_list, __pyx_v_i, Py_ssize_t, 1, PyInt_FromSsize_t, 0, 1, 1); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 10, __pyx_L1_error)
      __Pyx_GOTREF(__pyx_t_8);
      __pyx_t_9 = __Pyx_GetItemInt(__pyx_t_8, __pyx_v_j, Py_ssize_t, 1, PyInt_FromSsize_t, 0, 1, 1); if (unlikely(!__pyx_t_9)) __PYX_ERR(0, 10, __pyx_L1_error)
      __Pyx_GOTREF(__pyx_t_9);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
      __pyx_t_8 = PyNumber_InPlaceMultiply(__pyx_t_4, __pyx_t_9); if (unlikely(!__pyx_t_8)) __PYX_ERR(0, 10, __pyx_L1_error)
      __Pyx_GOTREF(__pyx_t_8);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
      __pyx_t_10 = __pyx_PyFloat_AsDouble(__pyx_t_8); if (unlikely((__pyx_t_10 == (double)-1) && PyErr_Occurred())) __PYX_ERR(0, 10, __pyx_L1_error)
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
      __pyx_v_count = __pyx_t_10;
```

Some of that is just indexing (and thus not very interesting). However:

Because `count` is a `float` and not an arbitrary Python object, it needs to be converted into a Python object, then have `PyNumber_InPlaceMultiply` called, then be converted back to a C float. This is because the type of the right-hand side of the multiplication isn't known and so Cython must fall back to generic Python multiplication.

If you hadn't typed `count` then all it would have to do is `PyNumber_InPlaceMultiply`.

**Comment by 00sapo** [comment]
@da-woods , thank you for the explanation. This actually explains why no typed is faster than typed code; even leaving lists to #3910, there's no clear reson why C++ is so much slower than C, nor why cython types are worst than pure python types
